pandas/core/apply.py

```diff
@@ -204,16 +204,19 @@ def apply_empty_result(self):
         from pandas import Series
 
         if not reduce:
-
-            EMPTY_SERIES = Series([])
             try:
-                r = self.f(EMPTY_SERIES, *self.args, **self.kwds)
+                r = self.f(Series([]))
```

Comment:
did you remove passing the args, kwargs on purpose? (I am not sure we are testing these)

Reply:
That was done on purpose since the the args were already passed here: https://github.com/pandas-dev/pandas/blob/cb68153f4b6fca19440ec6b79a0d1128c002ec11/pandas/core/apply.py#L112
This meant that for `nunique` (and possibly other functions) the line inside the `try` clause would raise a syntax error and `reduce` would never become `True`, and so we just end up returning a copy of the empty `DataFrame` due to the logic below.

Reply:
hmm, I am surpised this didn't break anything else, nor am I convinced this is the correct change (not that the original is any better); can you narrow the Exception here?

Reply:
FWIW ive got a branch that tries to narrow this exception.  IIRC part of the problem is that DataFrame.applymap defines

```
def infer(x):
    if x.empty:
        return lib.map_infer(x, func)
    return lib.map_infer(x.astype(object).values, func)
```

and in the empty case that raises because it `x` is a Series where map_infer expects an ndarray.  Then IIRC there was an issue that I fixed by making `infer` return `self._constructor(lib.map_infer(x._values, func), index=self.index, columns=self.columns)`, but that turned out to cause (hidden) problems elsewhere, and that was where that branch left off.

Reply:
If we remove _all_ error-catching here then two tests fail, the one @jbrockmendel mentions and `test_apply_empty`, which raises a `KeyError`.  This is because the test tries to apply the function `lambda x: x["a"]` over the rows of `DataFrame(index=["a"])`, but obviously this won't work because we're passing the barren `Series([])` object to that function in the `try` clause.  It works on `master` because the expected output is just the original `DataFrame` which gets returned down below.

So I guess something like `df.apply(lambda x: x.name)` would still give incorrect output on an empty `DataFrame` after this.

Reply:
IIUC this PR isn't intended for applymap, just apply, right?  i.e. the snippet from my previous comment is out of scope?

Reply:
> IIUC this PR isn't intended for applymap, just apply, right? i.e. the snippet from my previous comment is out of scope?

Originally I was just thinking `apply`, but if it affects `applymap` as well I'd say it's relevant

Reply:
Narrowed the exception down the the minimal set that still allows the tests to pass.  `ValueError` because of the `Series` vs. `ndarray` thing, `KeyError` because of the test that looks specifically at the column name.

Reply:
are there any restrictions on what `self.f` can be?  i.e. it can be a weird user-provided function right?  If so, then we probably can't get much more specific here.

Reply:
Yeah, I think it can be any arbitrary user-defined function

Reply:
We probably want to keep `Exception` then to not introduce an accidental behavior change.